base/linalg/diagonal.jl

```diff
@@ -150,9 +150,9 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, copy(B))
 
 (*)(A::AbstractMatrix, D::Diagonal) =
-    scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), A, D.diag)
+    scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), A, D.diag)
```

Comment:
This fix is a bit dirty; currently `similar(A::Symmetric)` return a `Matrix` iff you supply both a new element type and size.
Ref: #19016 #15198 and #13731